File: 2024_rust/src/day15_24.rs
use super::tools;
use std::time::Instant;
use std::collections::{HashSet, HashMap};
use itertools::Itertools;

fn get_object_ids_at_position(objects: &Vec<Vec<(i32, i32)>>, positions: &HashSet<(i32, i32)>) -> HashSet<usize> {
    positions.iter().fold(HashSet::<usize>::new(), |mut set, pos| {
        objects.iter().enumerate().for_each(|(i, obj)| {
            if obj.contains(pos) {
                set.insert(i);
            }
        });
        set
    })
}

fn can_move(map: &HashSet<(i32, i32)>, objects: &Vec<Vec<(i32, i32)>>, pos_set: &HashSet<(i32, i32)>, dir: &(i32, i32)) -> (bool, HashSet<usize>) {
    let next_pos_set = pos_set.iter().fold(HashSet::<(i32,i32)>::new(), |mut set, pos| {
        set.insert((pos.0 + dir.0, pos.1 + dir.1)); set
    });

    let obj_ids_in_front = get_object_ids_at_position(objects, &next_pos_set);

    let object_pos_in_front = obj_ids_in_front.iter().fold(HashSet::<(i32,i32)>::new(), |mut set, id| {
        objects[*id].iter().for_each(|&pos| {set.insert(pos);}); set
    });

    let next_object_pos_in_front = obj_ids_in_front.iter().fold(HashSet::<(i32,i32)>::new(), |mut set, id| {
        objects[*id].iter().for_each(|&pos| {set.insert((pos.0 + dir.0, pos.1 + dir.1 ) );}); set
    });

    let next_front: HashSet<(i32,i32)> = next_object_pos_in_front.difference(&object_pos_in_front)
        .into_iter()
        .map(|pos| (pos.0 - dir.0, pos.1 - dir.1))
        .fold(HashSet::<(i32,i32)>::new(), |mut set, pos| {set.insert(pos); set});

    if next_pos_set.difference(&next_front).all(|pos| map.contains(pos)) { // no walls in the way
        if obj_ids_in_front.is_empty() { // no objects in the way
            return (true, HashSet::new());
        }

        let result = can_move(map, objects, &next_front, dir);
        if result.0 {
            return (true, obj_ids_in_front.union(&result.1).cloned().collect());
        }
    }

    (false, HashSet::new())
}

fn do_operations(map: HashSet<(i32, i32)>, mut objects: Vec<Vec<(i32, i32)>>, start: (i32, i32), operations: &Vec<char>) -> usize {
    let dirs = vec![('^', (0, -1)), ('v', (0, 1)), ('<', (-1, 0)), ('>', (1, 0))]
        .into_iter()
        .collect::<HashMap<char, (i32, i32)>>();

    let mut robot = start;
    operations.iter().for_each(|op| {
        let dir = dirs.get(op).unwrap_or(&(0, 0));
        if dir == &(0, 0) {
            panic!("Invalid operation");
        }
        let (can, objs) = can_move(&map, &objects, &HashSet::from([robot;1]), dir);
        if can {
            objs.iter().for_each(|&i| {
                objects[i] = objects[i].iter().map(|pos| (pos.0 + dir.0, pos.1 + dir.1)).collect();
            });
            robot = (robot.0 + dir.0, robot.1 + dir.1);
        }
    });

    objects.iter().fold(0, |acc, obj| acc + obj[0].0 as usize + obj[0].1 as usize * 100)
}

#[allow(dead_code)]
pub fn run(real: bool, print_result: bool) -> (u128, u128, u128) {
    let start0 = Instant::now();

    let input_file: &str = if !real {
        "./input/day15-test.txt"
    } else {
        "./input/day15-real.txt"
    };
    let input = tools::get_input(String::from(input_file));
    let empty_line = input.iter().position(|r| r == "").unwrap();

    let mut map: HashSet<(i32, i32)> = HashSet::new();
    let mut objects: Vec<Vec<(i32, i32)>> = vec![];
    let mut start: (i32, i32) = (0, 0);
    input[..empty_line].iter().enumerate().for_each(|(y, line)| {
        line.chars().enumerate().for_each(|(x, c)| {
            let pos = (x as i32, y as i32);
            if c != '#' {
                map.insert(pos);
            }
            if c == 'O' {
                objects.push(vec![pos]);
            }
            if c == '@' {
                start = pos;
            }
        });
    });

    let wide_map: HashSet<(i32, i32)> = map.iter().fold(HashSet::<(i32,i32)>::new(), |mut set, pos| {
        set.insert((2 * pos.0 + 0, pos.1));
        set.insert((2 * pos.0 + 1, pos.1));
        set
    });

    let wide_objects: Vec<Vec<(i32, i32)>> = objects.iter().map(|obj| {
        vec![(2*obj[0].0+0, obj[0].1), (2*obj[0].0+1, obj[0].1)]
    }).collect(); 

    let wide_start = (2 * start.0, start.1);

    let operations = input[empty_line + 1..].iter().join("").chars().collect_vec();
    
    let after0 = Instant::now();

    let start1 = Instant::now();

    let res1 = do_operations(map.clone(), objects.clone(), start, &operations);

    let after1 = Instant::now();
    if print_result {
        println!("Part 1: {}", res1);
    }

    let start2 = Instant::now();

    let res2 = do_operations(wide_map.clone(), wide_objects.clone(), wide_start, &operations);

    let after2 = Instant::now();
    if print_result {
        println!("Part 2: {}", res2);
    }

    (
        after0.duration_since(start0).as_nanos(),
        after1.duration_since(start1).as_nanos(),
        after2.duration_since(start2).as_nanos(),
    )
}

File: 2024/day15-24/day15-24.py

# read command-line parameters and based on that read the input file
from collections import defaultdict
from copy import deepcopy
import sys
runtype = sys.argv[1]
runpart = int(sys.argv[2])
if len(sys.argv) > 3:
    runs = int(sys.argv[3])
else:
    runs = 1

text_file = open(f"day15-{runtype}.txt", "r")

lines = [line.strip() for line in text_file.readlines()]
emptyline = lines.index('')
operations = ''.join(lines[emptyline+1:])

size = (len(lines[0]), emptyline)
map = defaultdict(int)
objects = []
start = None

for y in range(size[1]):
    for x in range(size[0]):
        if lines[y][x] != '#':
            map[(x, y)] = 1
        if lines[y][x] == '@':
            start = (x, y)
        if lines[y][x] == 'O':
            objects.append([(x, y)])
            

def get_all_object_positions(objects, objects_to_do = None):
    objs = []
    for obj in objects:
        if objects_to_do is None or obj in objects_to_do:
            objs.extend(get_object_positions(obj))
    return objs
                
def get_object_positions(obj):
    return obj
    # if len(obj) == 1:
    #     objs = [obj[0]]
    # else:
    #     objs = [p for ob in obj for p in ob]
    # return objs
    
def print_map(map, objects, pos, size):
    objs = get_all_object_positions(objects)
    if len(objects[0]) == 1:
        objs = [obj[0] for obj in objects]
    else:
        objs = [p for obj in objects for p in obj]
    for y in range(size[1]):
        for x in range(size[0]):
            if (x, y) == pos:
                print('@', end='')
            elif (x, y) in objs:
                print('O', end='')
            elif (x,y) in map:
                print('.', end='')
            else:
                print('#', end='')
        print()
    print()    
    
def get_index_of_object(objects, pos):
    for i, obj in enumerate(objects):
        if pos in obj:
            return i
    return -1
    
    
def get_objects_at_position(objects, pos_set):
    return [obj for obj in objects if any([p in obj for p in pos_set])]

def get_object_ids_at_position(objects, pos_set):
    ids = []
    for i, obj in enumerate(objects):
        if any([p in obj for p in pos_set]):
            ids.append(i)
    return ids
    
def can_move(map, objects, pos_set, dir):
    next_pos_set = set([(pos[0] + dir[0], pos[1] + dir[1]) for pos in pos_set])
    
    obj_ids_in_front = set(get_object_ids_at_position(objects, next_pos_set))
    
    object_pos_in_front = set()
    for obj_id in obj_ids_in_front:
        object_pos_in_front = object_pos_in_front.union(set(objects[obj_id]))
    
    next_object_pos = set([(pos[0] + dir[0], pos[1] + dir[1]) for pos in object_pos_in_front])
    object_pos_in_front = next_object_pos.difference(object_pos_in_front)
    object_pos_in_front = set([(pos[0]-dir[0], pos[1]-dir[1]) for pos in object_pos_in_front])
    
    free_pos_in_front = next_pos_set.difference(object_pos_in_front)
    if all([pos in map for pos in free_pos_in_front]): # no walls in the way
        if len(obj_ids_in_front) == 0: # no objects in the way
            return True, set()
        
        result = can_move(map, objects, object_pos_in_front, dir)
        if result[0]:
            return True, obj_ids_in_front.union(result[1])
    
    return False, set()

def do_operations(map, objects, start, operations, size):
    op_index = 0
    robot = start
    dirs = {'^': (0, -1), '>': (1, 0), 'v': (0, 1), '<': (-1, 0)}    
    while op_index < len(operations):
        op = dirs[operations[op_index]]

        can, obj_ids = can_move(map, objects, set([robot]), op)
        if can:
            for index in obj_ids:
                for pi in range(len(objects[index])):
                    objects[index][pi] = (objects[index][pi][0] + op[0], objects[index][pi][1] + op[1])
            robot = (robot[0] + op[0], robot[1] + op[1])
        op_index += 1
    
    return sum([  100*obj[0][1] + obj[0][0] for obj in objects])

def part1():
    return do_operations(deepcopy(map), deepcopy(objects), start, operations, size)

def part2():
    widemap = defaultdict(int)
    for (x, y) in map.keys():
        widemap[(2*x+0, y)] = 1
        widemap[(2*x+1, y)] = 1
    wide_objects = deepcopy(objects)
    for i in range(len(wide_objects)):
        wide_objects[i] = [(2*wide_objects[i][0][0]+0, wide_objects[i][0][1]),
                           (2*wide_objects[i][0][0]+1, wide_objects[i][0][1])
                           ]
    wide_start = (2*start[0], start[1])
    wide_size = (2*size[0], size[1])
    
    # print_map(widemap, wide_objects, wide_start, wide_size)
    return do_operations(widemap, wide_objects, wide_start, operations, wide_size)

if runpart == 1 or runpart == 0:
    for run in range(runs):
        resp1 = part1()
    print("Part 1: {}".format(resp1))

if runpart == 2 or runpart == 0:
    for run in range(runs):
        resp2 = part2()
    print("Part 2: {}".format(resp2))
    
